pruning/keep_grid.go

```diff
@@ -82,16 +70,20 @@ func newKeepGrid(re *regexp.Regexp, configIntervals []config.RetentionInterval)
 	}
 
 	return &KeepGrid{
-		retentionGrid: retentiongrid.NewGrid(intervals),
-		re:            re,
+		common,
+		retentiongrid.NewGrid(intervals),
 	}, nil
 }
 
 // Prune filters snapshots with the retention grid.
 func (p *KeepGrid) KeepRule(snaps []Snapshot) (destroyList []Snapshot) {
 
 	matching, notMatching := partitionSnapList(snaps, func(snapshot Snapshot) bool {
-		return p.re.MatchString(snapshot.Name())
+		if p.negate {
```

Comment:
Please extract the `if p.negate  { ... } else { ... }` logic into `KeepCommon`. I see it repeated in every `KeepRule` implementation.
You can use a private (`lower-caps`) method on `KeepCommon` for that.